File: Plugins/GamePlugin/JavaGame/src/plugins/quorum/Libraries/Game/Graphics/IndexBufferObject.java
/*
 * To change this license header, choose License Headers in Project Properties.
 * To change this template file, choose Tools | Templates
 * and open the template in the editor.
 */
package plugins.quorum.Libraries.Game.Graphics;

import java.nio.Buffer;
import java.nio.IntBuffer;
import java.nio.ByteBuffer;
import plugins.quorum.Libraries.Game.GameRuntimeError;
import plugins.quorum.Libraries.Game.GameStateManager;
import plugins.quorum.Libraries.Game.libGDX.BufferUtils;
import quorum.Libraries.Containers.Integer32BitArray;
import quorum.Libraries.Containers.Integer32BitArray_;

/**
 *
 * @author alleew
 */
public class IndexBufferObject extends IndexData
{
    IntBuffer buffer;
    ByteBuffer byteBuffer;
    int bufferHandle;
    boolean isDirty = true;
    boolean isBound = false;
    int usage;
    
    private boolean empty;
    private int[] bridgeArray = null;
    
    public void Load(boolean isStatic, int maxIndices)
    {
        empty = maxIndices == 0;
        if (empty)
            maxIndices = 1; // Avoid allocating a zero-sized buffer due to a bug in Android's ART < Android 5.0
        
        byteBuffer = BufferUtils.newUnsafeByteBuffer(maxIndices * 4);
        
        buffer = byteBuffer.asIntBuffer();
        buffer.flip();
        byteBuffer.flip();
        bufferHandle = GameStateManager.nativeGraphics.glGenBuffer();
        usage = isStatic ? GraphicsManager.GL_STATIC_DRAW : GraphicsManager.GL_DYNAMIC_DRAW;
    }
    
    /** @return the number of indices currently stored in this buffer */
    public int GetSize()
    {
        return empty ? 0 : buffer.limit();
    }

    /** @return the maximum number of indices this IndexBufferObject can store. */
    public int GetMaxSize() 
    {
        return empty ? 0 : buffer.capacity();
    }

    @Override
    public void SetIndexRange(int start, int end)
    {
        buffer.position(start);
        buffer.limit(end);
    }
	
    public void SetIndices(int offset, int count) 
    {
        isDirty = true;
        buffer.clear();
        buffer.put(bridgeArray, offset, count);
        buffer.flip();
        byteBuffer.position(0);
        byteBuffer.limit(count << 1);

        if (isBound) 
        {
            GameStateManager.nativeGraphics.glBufferData(GraphicsManager.GL_ELEMENT_ARRAY_BUFFER, byteBuffer.limit(), byteBuffer, usage);
            isDirty = false;
        }
    }

    public void SetIndices (IntBuffer indices) 
    {
        isDirty = true;
        int pos = indices.position();
        buffer.clear();
        buffer.put(indices);
        buffer.flip();
        indices.position(pos);
        byteBuffer.position(0);
        byteBuffer.limit(buffer.limit() << 1);

        if (isBound) {
            GameStateManager.nativeGraphics.glBufferData(GraphicsManager.GL_ELEMENT_ARRAY_BUFFER, byteBuffer.limit(), byteBuffer, usage);
            isDirty = false;
        }
    }
    
    public IntBuffer GetBuffer()
    {
        isDirty = true;
        return buffer;
    }
    
    public void Clear()
    {
        buffer.clear();
    }
    
    /** Binds this IndexBufferObject for rendering with glDrawElements. */
    public void Bind()
    {
        if (bufferHandle == 0)
            throw new GameRuntimeError("Attempted to bind before a buffer was allocated!");

        GraphicsManager gl = GameStateManager.nativeGraphics;
        
        gl.glBindBuffer(GraphicsManager.GL_ELEMENT_ARRAY_BUFFER, bufferHandle);
        
        if (isDirty) 
        {
            byteBuffer.limit(buffer.limit() * 4);
            gl.glBufferData(GraphicsManager.GL_ELEMENT_ARRAY_BUFFER, byteBuffer.limit(), byteBuffer, usage);
            isDirty = false;
        }
        isBound = true;
    }

    /** Unbinds this IndexBufferObject. */
    public void Unbind() 
    {
        GameStateManager.nativeGraphics.glBindBuffer(GraphicsManager.GL_ELEMENT_ARRAY_BUFFER, 0);
        isBound = false;
    }

    /** Invalidates the IndexBufferObject so a new OpenGL buffer handle is created. Use this in case of a context loss. */
    public void Invalidate() 
    {
        bufferHandle = GameStateManager.nativeGraphics.glGenBuffer();
        isDirty = true;
    }

    /** Disposes this IndexBufferObject and all its associated OpenGL resources. */
    public void Dispose()
    {
        GameStateManager.nativeGraphics.glBindBuffer(GraphicsManager.GL_ELEMENT_ARRAY_BUFFER, 0);
        GameStateManager.nativeGraphics.glDeleteBuffer(bufferHandle);
        bufferHandle = 0;

        BufferUtils.disposeUnsafeByteBuffer(byteBuffer);
    }
    
    public void PrepareBridgeArray(int length)
    {
        bridgeArray = new int[length];
    }
    
    public void SendToBridgeArray(int index, int value)
    {
        bridgeArray[index] = value;
    }
    
    public void SendToBridgeArray(Integer32BitArray_ vertices)
    {
        Integer32BitArray array = (Integer32BitArray) vertices;
        plugins.quorum.Libraries.Containers.Integer32BitArray plugin = array.plugin_;
        bridgeArray = plugin.ints;
    }
    
    public void PutBridgeArray()
    {
        buffer.put(bridgeArray);
    }
    
    public void SetPosition(int position)
    {
        buffer.position(position);
    }
    
    @Override
    public void Reload()
    {
        bufferHandle = GameStateManager.nativeGraphics.glGenBuffer();
        isDirty = true;
    }
}

File: Quorum/Library/Standard/Plugins/javascript/Libraries/Game/Graphics/IndexBufferObject.js
function plugins_quorum_Libraries_Game_Graphics_IndexBufferObject_() 
{
    var buffer;
    var integerArray;
    var bridgeArray = [];
    var length = 0;
    var maxLength;
    var writingPosition = 0;
    
    var isDirty = true;
    var isBound = false;
    var usage;
    
    this.supports32bits;

    this.Load$quorum_boolean$quorum_integer = function(isStatic, maxIndices) 
    {
        var graphics = plugins_quorum_Libraries_Game_GameStateManager_.GetActiveGameInfo().nativeGraphics;
        
        this.supports32bits = graphics.gl.getExtension("OES_element_index_uint");
        
        maxLength = maxIndices;
        writingPosition = 0;
        
        buffer = graphics.glGenBuffer();
        
        if (isStatic)
            usage = graphics.gl.STATIC_DRAW;
        else
            usage = graphics.gl.DYNAMIC_DRAW;
    };
    
    this.GetSize = function() 
    {
        return length;
    };
    
    this.GetMaxSize = function() 
    {
        return maxLength;
    };
    
    this.SetIndices$quorum_integer$quorum_integer = function(offset, count) 
    {
        isDirty = true;
        
        if (length !== count)
        {
            this.SetLength(count);
            integerArray = this.NewIntegerArray(count);
        }
        
        for (var i = 0; i < count; i++)
            integerArray[i] = bridgeArray[offset + i];
        
        writingPosition = 0;
        
        if (isBound)
        {
            var graphics = plugins_quorum_Libraries_Game_GameStateManager_.GetActiveGameInfo().nativeGraphics;
            graphics.glBufferData(graphics.gl.ELEMENT_ARRAY_BUFFER, integerArray, usage);
            isDirty = false;
        }
    };
    
    this.Bind = function() 
    {
        if (buffer === undefined || buffer === null)
        {
            var exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("Attempted to bind before a buffer was allocated!");
            throw exceptionInstance_;  
        }
        
        var graphics = plugins_quorum_Libraries_Game_GameStateManager_.GetActiveGameInfo().nativeGraphics;
        
        graphics.glBindBuffer(graphics.gl.ELEMENT_ARRAY_BUFFER, buffer);
        
        if (isDirty)
        {
            graphics.glBufferData(graphics.gl.ELEMENT_ARRAY_BUFFER, integerArray, usage);
            isDirty = false;
        }
        isBound = true;
    };
    
    this.Unbind = function() 
    {
        var graphics = plugins_quorum_Libraries_Game_GameStateManager_.GetActiveGameInfo().nativeGraphics;
        graphics.glBindBuffer(graphics.gl.ELEMENT_ARRAY_BUFFER, null);
        isBound = false;
    };
    
    this.Invalidate = function() 
    {
        var graphics = plugins_quorum_Libraries_Game_GameStateManager_.GetActiveGameInfo().nativeGraphics;
        buffer = graphics.glGenBuffer();
        isDirty = true;
    };
    
    this.Clear = function() 
    {
        integerArray = null;
        length = 0;
        writingPosition = 0;
    };
    
    this.Reload = function()
    {
        // Do nothing. There's (currently) nothing to be done here.
    };
    
    this.Dispose = function() 
    {
        var graphics = plugins_quorum_Libraries_Game_GameStateManager_.GetActiveGameInfo().nativeGraphics;
        graphics.glBindBuffer(graphics.gl.ELEMENT_ARRAY_BUFFER, null);
        graphics.glDeleteBuffer(buffer);
        buffer = null;
        
        integerArray = null;
        bridgeArray = [];
        length = 0;
        maxLength = 0;
        writingPosition = 0;
    };
    
    this.PrepareBridgeArray$quorum_integer = function(newLength) 
    {
        if (length !== newLength)
        {
            this.SetLength(newLength);
            integerArray = this.NewIntegerArray(newLength);
        }
    };
    
    this.SendToBridgeArray$quorum_Libraries_Containers_Integer32BitArray = function(int32array)
    {
        bridgeArray = int32array.plugin_.array_;
    };
    
    this.PutBridgeArray = function() 
    {
        isDirty = true;
        
        if (length < writingPosition + bridgeArray.length)
        {
            this.SetLength(writingPosition + bridgeArray.length);
            integerArray = this.NewIntegerArray(writingPosition + bridgeArray.length);
        }
        
        for (var i = 0; i < bridgeArray.length; i++)
            integerArray[writingPosition + i] = bridgeArray[i];
        
        writingPosition = writingPosition + bridgeArray.length;
    };
    
    this.SetPosition$quorum_integer = function(position) 
    {
        writingPosition = position;
    };
    
    this.NewIntegerArray = function(size)
    {
        if (this.supports32bits)
            return new Uint32Array(size);
        else
            return new Uint16Array(size);
    };
    
    this.SetLength = function(newLength)
    {
        if (newLength > maxLength)
        {
            var exceptionInstance_ = new quorum_Libraries_Language_Errors_Error_();
            exceptionInstance_.SetErrorMessage$quorum_text("The IndexArray was asked to set its length to " + newLength + ", but the maximum length is " + maxLength);
            throw exceptionInstance_;  
        }
        
        length = newLength;
    };
    
    this.GetBuffer = function()
    {
        isDirty = true;
        return integerArray;
    };
}

/*
    private system action SendToBridgeArray(Integer32BitArray array)
 */